File: HikvisionService/Controllers/FootageController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Authorization;
using HikvisionService.Models;
using HikvisionService.Models.ViewModels;
using HikvisionService.Services;

namespace HikvisionService.Controllers;

[Authorize]
public class FootageController : Controller
{
    private readonly IHikvisionService _hikvisionService;
    private readonly ILogger<FootageController> _logger;

    public FootageController(IHikvisionService hikvisionService, ILogger<FootageController> logger)
    {
        _hikvisionService = hikvisionService;
        _logger = logger;
    }

    public async Task<IActionResult> Index(FootageViewModel model)
    {
        try
        {
            // Set default dates if not provided
            if (model.StartDate == default)
            {
                model.StartDate = DateTime.Today.AddDays(-7);
            }
            
            if (model.EndDate == default)
            {
                model.EndDate = DateTime.Today;
            }
            
            // Get all cameras for the dropdown
            model.Cameras = await _hikvisionService.GetAllCamerasAsync();
            
            // Get footage files based on filters
            model.Files = await _hikvisionService.GetFootageFilesAsync(
                model.SelectedCameraId,
                model.StartDate,
                model.EndDate,
                model.FileType
            );
            
            return View(model);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error loading footage data");
            return View(new FootageViewModel());
        }
    }
    
    [HttpGet]
    public async Task<IActionResult> Download(string path)
    {
        try
        {
            // Validate and get the file path
            var filePath = await _hikvisionService.GetFootageDownloadUrlAsync(path);
            
            // Get file info
            var fileInfo = new System.IO.FileInfo(filePath);
            if (!fileInfo.Exists)
            {
                return NotFound();
            }
            
            // Determine content type
            string contentType = "application/octet-stream";
            if (filePath.EndsWith(".mp4", StringComparison.OrdinalIgnoreCase))
            {
                contentType = "video/mp4";
            }
            else if (filePath.EndsWith(".jpg", StringComparison.OrdinalIgnoreCase) || 
                     filePath.EndsWith(".jpeg", StringComparison.OrdinalIgnoreCase))
            {
                contentType = "image/jpeg";
            }
            
            // Return the file
            return PhysicalFile(filePath, contentType, fileInfo.Name);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error downloading file: {Path}", path);
            return BadRequest("Error downloading file");
        }
    }
}

File: laravel/app/Http/Controllers/FootageController.php
<?php

namespace App\Http\Controllers;

use App\Models\Camera;
use App\Models\FileDownloadJob;
use App\Services\HikvisionService;
use Carbon\Carbon;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

class FootageController extends Controller
{
    protected $hikvisionService;

    public function __construct(HikvisionService $hikvisionService)
    {
        $this->hikvisionService = $hikvisionService;
    }

    public function index(Request $request)
    {
        try {
            $cameras = Camera::orderBy('name')->get();

            // Default to last 7 days
            $startDate = $request->input('start_date', now()->subDays(7)->format('Y-m-d'));
            $endDate = $request->input('end_date', now()->format('Y-m-d'));
            $cameraId = $request->input('camera_id');
            $fileType = $request->input('file_type', 'both');

            $footageFiles = [];
            if ($request->has('search')) {
                $footageFiles = $this->hikvisionService->getFootageFiles(
                    $cameraId,
                    Carbon::parse($startDate),
                    Carbon::parse($endDate),
                    $fileType
                );
            }

            return view('footage.index', compact(
                'cameras',
                'footageFiles',
                'startDate',
                'endDate',
                'cameraId',
                'fileType'
            ));
        } catch (\Exception $ex) {
            Log::error('Error getting footage files', ['error' => $ex->getMessage()]);
            return redirect()->back()->with('error', 'An error occurred while retrieving footage files.');
        }
    }

    public function download(Request $request)
    {
        try {
            $filePath = $request->query('path');

            if (!$filePath) {
                return redirect()->back()->with('error', 'File path is required.');
            }

            $downloadUrl = $this->hikvisionService->getFootageDownloadUrl($filePath);

            if (!file_exists($downloadUrl)) {
                return redirect()->back()->with('error', 'File not found.');
            }

            // Determine content type
            $contentType = 'application/octet-stream';
            if (str_ends_with(strtolower($filePath), '.mp4')) {
                $contentType = 'video/mp4';
            } elseif (str_ends_with(strtolower($filePath), '.jpg') || str_ends_with(strtolower($filePath), '.jpeg')) {
                $contentType = 'image/jpeg';
            }

            return response()->file($downloadUrl, [
                'Content-Type' => $contentType,
                'Content-Disposition' => 'attachment; filename="' . basename($filePath) . '"',
            ]);
        } catch (\Exception $ex) {
            Log::error("Error downloading footage file: {$request->query('path')}", ['error' => $ex->getMessage()]);
            return redirect()->back()->with('error', 'Failed to download file.');
        }
    }
}
